mlir/lib/Conversion/EmulateFp8ExtTrunc/EmulateFp8ExtTrunc.cpp

```diff
@@ -119,7 +127,18 @@ static Value getFloatValueTableFor(Type elementType, Operation *op,
   const auto &sem = type.getFloatSemantics();
   for (uint32_t i = 0; i < 256; ++i) {
     APFloat entry(sem, APInt(8, i));
-    tableElems.push_back(entry.convertToFloat());
+    float x = entry.convertToFloat();
+    uint32_t u = llvm::bit_cast<uint32_t>(x);
+    // Ugly hack:  APFloat will turn signaling NaN into quiet Nan, so to get
+    // correct reversible expansion we force this special case.
+    if (type.isFloat8E5M2()) {
+      if (i == 0x7d)
+        u = 0x7fa00000;
+      if (i == 0xfd)
+        u = 0xffa00000;
```

Comment:
Do we know that these are supposed to be qnans?

Reply:
I went with the idea that the mantissa bits in 8-bit should become the mantissa bits in 32-bit, and for NaN the exponent is all-ones in both.  In OCP e5m2, the NaNs are 0x7d, 0x7e, and 0x7f, which would then become 0x7fa00000, 0x7fc00000, and 0x7fe0000.  APFloat was forcing 0x7fa00000 to 0x7fe00000, when meant that 0x7d, extended and then truncated, came back as 0x7f which failed the roundtrip.mlir test.

Reply:
I'd go double-check with hardware / anyone else's details on what's supposed to happen here